internal/cluster-object/service_instance_provider.go

```diff
@@ -0,0 +1,86 @@
+package clusterobject
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	"github.com/kyma-project/btp-manager/controllers"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	serviceInstanceProviderName = "ServiceInstanceProvider"
+	secretRefKey                = "btpAccessCredentialsSecret"
+)
+
+type ServiceInstanceProvider struct {
+	client.Reader
+	logger *slog.Logger
+}
+
+func NewServiceInstanceProvider(reader client.Reader, logger *slog.Logger) *ServiceInstanceProvider {
+	logger = logger.With(logComponentNameKey, serviceInstanceProviderName)
+
+	return &ServiceInstanceProvider{
+		Reader: reader,
+		logger: logger,
+	}
+}
+
+func (p *ServiceInstanceProvider) AllWithSecretRef(ctx context.Context) (*unstructured.UnstructuredList, error) {
+	filtered, err := p.All(ctx)
```

Comment:
Could we use Field Selectors for loading instances that contains secret refs directly from K8S Api?

Reply:
I wish, field selector is very limited, see: https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/#supported-fields